content/blog/end-to-end-easy-to-use-pipeline-for-training-a-model-on-medmnist-v2-using-hpe-machine-learning-development-environment-flask.md

```diff
@@ -0,0 +1,514 @@
+
+```python
+resources:
+    slots_per_trial: 4
+```
+
+After taking these steps, you’d be able to watch your experiment progress in the WebUI and be assured that you’re utilizing multiple GPUs. By viewing the “Cluster” tab, you will notice that 4/5 CUDA Slots are allocated and in use: 
+
+![](/img/screenshot2.png)
+
+#### Automatic Checkpointing 
+
+Checkpointing a model throughout an experiment is important to maintain training progress and for preserving the best model at the end of an experiment. Let’s take a look at how the original training script handles model checkpointing: 
+
+The original training script saves the last model at the very end of the training process, in a location specified by the user, but does not checkpoint throughout the training job: 
+
+```python
+    path = os.path.join(output_root, 'best_model.pth')
+    torch.save(state, path)
+```
+
+This approach is problematic for the following reasons: 
+
+* The last model, at the end of a training job, is not always the best model. 
+* The script could fail before the end of a training job, resulting in no model checkpoint, which wastes time and resources. 
+
+Determined automatically checkpoints in the following situations: 
+
+* Periodically throughout the course of model training, to keep a record of the training progress. 
+* During training – If a trial fails on epoch 9 and the last checkpoint was saved during epoch 1, Determined will save yet another checkpoint at epoch 9, making this a very efficient system. 
+* Upon completion of the trial. 
+
+Additional checkpoint configuration settings can be modified to make this even more customizable. Checkpoints can be easily examined through the WebUI even after an experiment completes, and downloaded easily through the Determined APIs: 
+
+![](/img/screenshot3.png)
+
+#### Hyperparameter search 
+
+Hyperparameter search refers to the process of searching for the optimal configuration settings for your machine learning model – for example, searching for the optimal learning rate or convolution sizes in a convolutional neural network. The original training script does not implement hyperparameter search. This is not surprising, as hyperparameter search is yet another heavy lift for a researcher who wants to focus on experimentation, not the grunt work of setting up hyperparameter search.  
+
+With Determined, configuring hyperparameter search is easy. Defining hyperparameters, either static or in ranges, is easy through experiment configuration: 
+
+```yaml
+hyperparameters:
+    global_batch_size: 128
+    data_flag: pathmnist
+    dataset_name: "pathmnist.npz"
+    model_flag: resnet18
+    lr: 0.001
+    gamma: 0.1
+    resize: True
+    task: "multi-class"
+    num_epochs: 15
+```
+
+In this case, the “hyperparameters” feature can also be used to switch between datasets and types of training tasks if configured appropriately in the training script – which is neat! 
+
+#### Collaboration 
```

Comment:
If "Resource management" and "Model sharing and reproducibility" are both types of "Collaboration, then the header for "Collaboration" should be a size larger.

Reply:
Great catch